File: packages/go/tx_test.go
package rdk

import "testing"

func TestSignVerifyRoundtrip(t *testing.T) {
	g := loadGolden(t)
	acc, err := DeriveNativeAccount(g.Mnemonic, 0)
	if err != nil {
		t.Fatal(err)
	}

	msgs := []Msg{MsgCreateRollup{
		Creator: acc.Address, RollupID: "r1", Profile: "defi", VmType: "evm", StakeAmount: 10000000000,
	}}
	fee := Fee{Amount: []Coin{{Denom: "uqor", Amount: "2000"}}, GasLimit: 200000}

	signDoc := SignDocBytes(msgs, "", fee, acc.PublicKey, 0, "qorechain-diana", 0)
	sig := SignSignDoc(signDoc, acc.PrivateKey)
	if len(sig) != 64 {
		t.Fatalf("expected 64-byte signature, got %d", len(sig))
	}
	if !VerifySignature(signDoc, sig, acc.PublicKey) {
		t.Error("signature failed to verify against pubkey")
	}

	// A different message must not verify against the same signature.
	otherDoc := SignDocBytes(msgs, "tampered", fee, acc.PublicKey, 0, "qorechain-diana", 0)
	if VerifySignature(otherDoc, sig, acc.PublicKey) {
		t.Error("signature verified against a different sign doc")
	}
}

func TestSignTxProducesTxRaw(t *testing.T) {
	g := loadGolden(t)
	acc, _ := DeriveNativeAccount(g.Mnemonic, 0)
	msgs := []Msg{MsgPauseRollup{Creator: acc.Address, RollupID: "r1", Reason: "maintenance"}}
	fee := Fee{Amount: []Coin{{Denom: "uqor", Amount: "1000"}}, GasLimit: 100000}
	raw := SignTx(acc, msgs, "", fee, 5, "qorechain-vladi", 12)
	if len(raw) == 0 {
		t.Fatal("TxRaw should not be empty")
	}
	// TxRaw begins with field 1 (body_bytes), tag 0x0a.
	if raw[0] != 0x0a {
		t.Errorf("TxRaw should start with body_bytes field tag 0x0a, got %#x", raw[0])
	}
}

func TestEncodeAnyShape(t *testing.T) {
	any := encodeAny("/x.Y", []byte{0x01, 0x02})
	// field1 string "/x.Y": 0a 04 2f782e59 ; field2 bytes 0102: 12 02 0102
	want := "0a042f782e5912020102"
	if BytesToHex(any) != want {
		t.Errorf("encodeAny got %s want %s", BytesToHex(any), want)
	}
}

File: packages/jvm/src/test/java/io/github/qorechain/rdk/TxTest.java
package io.github.qorechain.rdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.github.qorechain.rdk.accounts.Account;
import io.github.qorechain.rdk.accounts.Accounts;
import io.github.qorechain.rdk.tx.Messages;
import io.github.qorechain.rdk.tx.Messages.MsgCreateRollup;
import io.github.qorechain.rdk.tx.Messages.MsgExecuteWithdrawal;
import io.github.qorechain.rdk.tx.Messages.MsgPauseRollup;
import io.github.qorechain.rdk.tx.Messages.MsgSubmitBatch;
import io.github.qorechain.rdk.tx.Msg;
import io.github.qorechain.rdk.tx.Tx;
import io.github.qorechain.rdk.util.Bytes;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

class TxTest {

    @Test
    void msgProtoBytesMatchGolden() {
        Golden g = Golden.load();

        MsgCreateRollup createRollup = new MsgCreateRollup();
        createRollup.creator = "qor1creator";
        createRollup.rollupId = "my-rollup";
        createRollup.profile = "defi";
        createRollup.vmType = "evm";
        createRollup.stakeAmount = 10000000000L;

        MsgPauseRollup pauseRollup = new MsgPauseRollup();
        pauseRollup.creator = "qor1creator";
        pauseRollup.rollupId = "r1";
        pauseRollup.reason = "x";

        MsgSubmitBatch submitBatch = new MsgSubmitBatch();
        submitBatch.sequencer = "qor1seq";
        submitBatch.rollupId = "r";
        submitBatch.batchIndex = 7;
        submitBatch.stateRoot = new byte[] {0x01, 0x02, 0x03};
        submitBatch.prevStateRoot = new byte[] {0x04, 0x05};
        submitBatch.txCount = 42;
        submitBatch.dataHash = new byte[] {0x09, 0x09};
        submitBatch.proof = new byte[] {0x08};
        submitBatch.withdrawalsRoot = new byte[] {0x07, 0x07, 0x07};

        MsgExecuteWithdrawal executeWithdrawal = new MsgExecuteWithdrawal();
        executeWithdrawal.submitter = "qor1sub";
        executeWithdrawal.rollupId = "r";
        executeWithdrawal.batchIndex = 3;
        executeWithdrawal.withdrawalIndex = 1;
        executeWithdrawal.recipient = "qor1dest";
        executeWithdrawal.denom = "uqor";
        executeWithdrawal.amount = 500;
        List<byte[]> proof = new ArrayList<>();
        proof.add(new byte[] {0x01});
        proof.add(new byte[] {0x02, 0x02});
        executeWithdrawal.proof = proof;

        assertEquals(g.msgProtoHex.get("MsgCreateRollup"), Bytes.bytesToHex(createRollup.marshal()));
        assertEquals(g.msgProtoHex.get("MsgPauseRollup"), Bytes.bytesToHex(pauseRollup.marshal()));
        assertEquals(g.msgProtoHex.get("MsgSubmitBatch"), Bytes.bytesToHex(submitBatch.marshal()));
        assertEquals(
                g.msgProtoHex.get("MsgExecuteWithdrawal"),
                Bytes.bytesToHex(executeWithdrawal.marshal()));
    }

    @Test
    void typeUrlsMatch() {
        assertEquals("/qorechain.rdk.v1.MsgCreateRollup", Messages.TYPE_URL_MSG_CREATE_ROLLUP);
        assertEquals("/qorechain.rdk.v1.MsgSubmitBatch", new MsgSubmitBatch().typeUrl());
        assertEquals("/qorechain.rdk.v1.MsgExecuteWithdrawal", new MsgExecuteWithdrawal().typeUrl());
    }

    @Test
    void signSignDocRoundTripsThroughVerify() {
        Golden g = Golden.load();
        Account acc = Accounts.deriveNativeAccount(g.mnemonic, 0);

        MsgPauseRollup pause = new MsgPauseRollup();
        pause.creator = acc.address;
        pause.rollupId = "r1";
        pause.reason = "maintenance";

        Tx.Fee fee =
                new Tx.Fee(List.of(new Tx.Coin("uqor", "2000")), 120000);
        byte[] signDoc =
                Tx.signDocBytes(List.<Msg>of(pause), "", fee, acc.publicKey, 0, "qorechain-diana", 0);

        byte[] sig = Tx.signSignDoc(signDoc, acc.privateKey);
        assertEquals(64, sig.length);
        assertTrue(Tx.verifySignature(signDoc, sig, acc.publicKey));

        // A tampered message must not verify.
        byte[] tampered = signDoc.clone();
        tampered[tampered.length - 1] ^= 0x01;
        assertFalse(Tx.verifySignature(tampered, sig, acc.publicKey));
    }

    @Test
    void signTxProducesNonEmptyTxRaw() {
        Golden g = Golden.load();
        Account acc = Accounts.deriveNativeAccount(g.mnemonic, 0);
        MsgPauseRollup pause = new MsgPauseRollup();
        pause.creator = acc.address;
        pause.rollupId = "r1";
        Tx.Fee fee = new Tx.Fee(List.of(new Tx.Coin("uqor", "2000")), 120000);
        byte[] txRaw = Tx.signTx(acc, List.<Msg>of(pause), "", fee, 0, "qorechain-diana", 0);
        assertTrue(txRaw.length > 0);
    }
}
